custom/icds_reports/utils/aggregation_helpers/distributed/agg_dashboard_activity.py

```diff
@@ -0,0 +1,276 @@
+                      for index, loc in enumerate(self.get_user_locations())
+                      }
+        param_keys = ['%({})s'.format(param) for param in parameters.keys()]
+
+        return """
+        INSERT INTO "{tablename}" (
+            username, state_id, district_id,block_id,
+            user_level,date
+        )
+        VALUES {param_keys}
+        """.format(
+            tablename=self.tablename,
+            param_keys=','.join(param_keys)
+        ), parameters
+
+    def rollover_previous_data(self):
+        query_param = {'date': self.date,
+                       'last_agg_date': self.last_agg_date}
+
+        return """
+        UPDATE "{tablename}" user_activity SET
+            location_launched = ut.location_launched,
+            last_activity = ut.last_activity
+        FROM  (
+        SELECT
+            username,
+            location_launched,
+            last_activity
+            FROM "{parent_tablename}" WHERE date = %(last_agg_date)s
+            )ut
+        WHERE user_activity.username = ut.username
+        """.format(
+            tablename=self.tablename,
+            parent_tablename=self.aggregate_parent_table
+        ), query_param
+
+    def update_queries(self):
+        last_time_to_consider = self.date
+        latest_month = (last_time_to_consider - timedelta(days=1)).replace(day=1)
+
+        yield """
+        UPDATE "{tablename}" user_activity
+            SET location_launched = CASE WHEN num_launched_blocks>0 THEN TRUE ELSE FALSE END
+        FROM (
+            SELECT
+                block_id, district_id, state_id, num_launched_blocks,aggregation_level
+            FROM agg_awc where month=%(latest_month)s and aggregation_level<=3
+            ) ut
+        WHERE (
+            user_activity.user_level= ut.aggregation_level AND
+            user_activity.state_id = ut.state_id AND
+            user_activity.district_id = ut.district_id AND
+            user_activity.block_id = ut.block_id AND
+            user_activity.location_launched is not TRUE
+        )
+        """.format(
+            tablename=self.tablename
+        ), {
+            'latest_month': latest_month
+        }
+
+        # This is query I prepare which could do what the last two queries are doing but
+        # the cost of this single query coming out to be very high
+        # yield """
+        # UPDATE {tablename} user_activity
+        # SET
+        #     last_activity
+        # from (
+        # SELECT audit.username, max(audit.time_of_use) from {tablename} user_activity left join
+        # icds_audit_entry_record audit
+        # on user_activity.username = audit.username
+        # group by username where (user_activity.last_activity is not null AND
+        #                          time_of_use>user_activity.last_activity
+        #                           ) or
+        #                         ( user_activity.last_activity is null)
```

Comment:
Could this query be `where icds_audit_entry_record >= last_agg_date` and in the select `GREATEST(audit.time_of_use, user_activity.last_activity)`?

ORs are generally expensive in sql, and it would be good to get rid of the next 2 queries

Reply:
I don't think it can be applied with `where icds_audit_entry_record >= last_agg_date`. Because for the first time at least the query has to search the icds_audit_entry_record for entire history to fetch the last activity time of user because the first user_activity table will have null for all users.

Reply:
So you're only doing this because of the inital filling query? As I mentioned in an above comment, I think if your helper just returns an old date, teh initial query hsould "just work" right?

Reply:
hey, I think that would work in a single query. But I did a simple analysis with a dummy report of three users and cost of that single query came very high. lot more than below two queries combined.

I tried this query:

```
EXPLAIN UPDATE  temp_rohit_activity user_activity
        SET
            last_activity = ut.last_activity
        FROM (
        SELECT audit.username, max(audit.time_of_use)  as last_activity from temp_rohit_activity left join icds_audit_entry_record audit
        oN temp_rohit_activity.username = audit.username
        where audit.time_of_use>'1970-01-01' and audit.time_of_use<'2019-12-09'
        group by audit.username
        )ut 
        where user_activity.username = ut.username;
```
It gave below cost. 

```
Update on temp_rohit_activity user_activity  (cost=4316146.33..4316257.19 rows=34254 width=113)
   ->  Hash Join  (cost=4316146.33..4316257.19 rows=34254 width=113)
         Hash Cond: ((user_activity.username)::text = (ut.username)::text)
         ->  Seq Scan on temp_rohit_activity user_activity  (cost=0.00..22.00 rows=1200 width=38)
         ->  Hash  (cost=4316074.97..4316074.97 rows=5709 width=110)
               ->  Subquery Scan on ut  (cost=4315960.79..4316074.97 rows=5709 width=110)
                     ->  HashAggregate  (cost=4315960.79..4316017.88 rows=5709 width=43)
                           Group Key: audit.username
                           ->  Nested Loop  (cost=0.56..4263691.98 rows=10453763 width=43)
                                 ->  Seq Scan on temp_rohit_activity  (cost=0.00..22.00 rows=1200 width=32)
                                 ->  Append  (cost=0.56..3286.06 rows=26700 width=43)
                                       ->  Index Scan using icds_audit_entry_record_username_ed335fe7 on icds_audit_entry_record audit  (cost=0.56..767.09 rows=2470 width=43)
                                             Index Cond: ((username)::text = (temp_rohit_activity.username)::text)
                                             Filter: ((time_of_use > '1970-01-01 00:00:00+00'::timestamp with time zone) AND (time_of_use < '2019-12-09 00:00:00+00'::timestamp with time zone))
                                       ->  Index Scan using icds_audit_entry_record_y2019m06_username_idx on icds_audit_entry_record_y2019m06 audit_1  (cost=0.43..106.13 rows=796 width=43)
                                             Index Cond: ((username)::text = (temp_rohit_activity.username)::text)
                                             Filter: ((time_of_use > '1970-01-01 00:00:00+00'::timestamp with time zone) AND (time_of_use < '2019-12-09 00:00:00+00'::timestamp with time zone))
                                       ->  Index Scan using icds_audit_entry_record_y2018m11_username_idx on icds_audit_entry_record_y2018m11 audit_2  (cost=0.55..138.70 rows=2646 width=43)
                                             Index Cond: ((username)::text = (temp_rohit_activity.username)::text)
                                             Filter: ((time_of_use > '1970-01-01 00:00:00+00'::timestamp with time zone) AND (time_of_use < '2019-12-09 00:00:00+00'::timestamp with time zone))
                                       ->  Index Scan using icds_audit_entry_record_y2018m10_username_idx on icds_audit_entry_record_y2018m10 audit_3  (cost=0.42..12.91 rows=246 width=43)
                                             Index Cond: ((username)::text = (temp_rohit_activity.username)::text)
                                             Filter: ((time_of_use > '1970-01-01 00:00:00+00'::timestamp with time zone) AND (time_of_use < '2019-12-09 00:00:00+00'::timestamp with time zone))
                                       ->  Index Scan using icds_audit_entry_record_y2018m05_username_idx1 on icds_audit_entry_record_y2018m05 audit_4  (cost=0.42..12.50 rows=244 width=44)
                                             Index Cond: ((username)::text = (temp_rohit_activity.username)::text)
                                             Filter: ((time_of_use > '1970-01-01 00:00:00+00'::timestamp with time zone) AND (time_of_use < '2019-12-09 00:00:00+00'::timestamp with time zone))
                                       ->  Index Scan using icds_audit_entry_record_y2018m06_username_idx on icds_audit_entry_record_y2018m06 audit_5  (cost=0.42..45.50 rows=735 width=43)
                                             Index Cond: ((username)::text = (temp_rohit_activity.username)::text)
                                             Filter: ((time_of_use > '1970-01-01 00:00:00+00'::timestamp with time zone) AND (time_of_use < '2019-12-09 00:00:00+00'::timestamp with time zone))
                                       ->  Index Scan using icds_audit_entry_record_y2018m07_username_idx on icds_audit_entry_record_y2018m07 audit_6  (cost=0.42..68.53 rows=1040 width=43)
                                             Index Cond: ((username)::text = (temp_rohit_activity.username)::text)
                                             Filter: ((time_of_use > '1970-01-01 00:00:00+00'::timestamp with time zone) AND (time_of_use < '2019-12-09 00:00:00+00'::timestamp with time zone))
                                       ->  Index Scan using icds_audit_entry_record_y2018m08_username_idx on icds_audit_entry_record_y2018m08 audit_7  (cost=0.56..286.06 rows=5146 width=43)
                                             Index Cond: ((username)::text = (temp_rohit_activity.username)::text)
                                             Filter: ((time_of_use > '1970-01-01 00:00:00+00'::timestamp with time zone) AND (time_of_use < '2019-12-09 00:00:00+00'::timestamp with time zone))
                                       ->  Index Scan using icds_audit_entry_record_y2019m03_username_idx on icds_audit_entry_record_y2019m03 audit_8  (cost=0.43..67.15 rows=883 width=43)
                                             Index Cond: ((username)::text = (temp_rohit_activity.username)::text)
                                             Filter: ((time_of_use > '1970-01-01 00:00:00+00'::timestamp with time zone) AND (time_of_use < '2019-12-09 00:00:00+00'::timestamp with time zone))
                                       ->  Index Scan using icds_audit_entry_record_y2019m04_username_idx on icds_audit_entry_record_y2019m04 audit_9  (cost=0.43..106.56 rows=1440 width=43)
                                             Index Cond: ((username)::text = (temp_rohit_activity.username)::text)
                                             Filter: ((time_of_use > '1970-01-01 00:00:00+00'::timestamp with time zone) AND (time_of_use < '2019-12-09 00:00:00+00'::timestamp with time zone))
                                       ->  Index Scan using icds_audit_entry_record_y2019m01_username_idx on icds_audit_entry_record_y2019m01 audit_10  (cost=0.43..112.06 rows=1958 width=42)
                                             Index Cond: ((username)::text = (temp_rohit_activity.username)::text)
                                             Filter: ((time_of_use > '1970-01-01 00:00:00+00'::timestamp with time zone) AND (time_of_use < '2019-12-09 00:00:00+00'::timestamp with time zone))
                                       ->  Index Scan using icds_audit_entry_record_y2018m12_username_idx on icds_audit_entry_record_y2018m12 audit_11  (cost=0.43..80.69 rows=1449 width=43)
                                             Index Cond: ((username)::text = (temp_rohit_activity.username)::text)
                                             Filter: ((time_of_use > '1970-01-01 00:00:00+00'::timestamp with time zone) AND (time_of_use < '2019-12-09 00:00:00+00'::timestamp with time zone))
                                       ->  Index Scan using icds_audit_entry_record_y2019m02_username_idx on icds_audit_entry_record_y2019m02 audit_12  (cost=0.56..153.69 rows=2065 width=43)
                                             Index Cond: ((username)::text = (temp_rohit_activity.username)::text)
                                             Filter: ((time_of_use > '1970-01-01 00:00:00+00'::timestamp with time zone) AND (time_of_use < '2019-12-09 00:00:00+00'::timestamp with time zone))
                                       ->  Index Scan using icds_audit_entry_record_y2018m09_username_idx on icds_audit_entry_record_y2018m09 audit_13  (cost=0.42..27.30 rows=528 width=43)
                                             Index Cond: ((username)::text = (temp_rohit_activity.username)::text)
                                             Filter: ((time_of_use > '1970-01-01 00:00:00+00'::timestamp with time zone) AND (time_of_use < '2019-12-09 00:00:00+00'::timestamp with time zone))
                                       ->  Index Scan using icds_audit_entry_record_y2019m05_username_idx on icds_audit_entry_record_y2019m05 audit_14  (cost=0.56..169.88 rows=1599 width=43)
                                             Index Cond: ((username)::text = (temp_rohit_activity.username)::text)
                                             Filter: ((time_of_use > '1970-01-01 00:00:00+00'::timestamp with time zone) AND (time_of_use < '2019-12-09 00:00:00+00'::timestamp with time zone))
                                       ->  Index Scan using icds_audit_entry_record_y2019m10_username_idx on icds_audit_entry_record_y2019m10 audit_15  (cost=0.43..91.23 rows=363 width=44)
                                             Index Cond: ((username)::text = (temp_rohit_activity.username)::text)
                                             Filter: ((time_of_use > '1970-01-01 00:00:00+00'::timestamp with time zone) AND (time_of_use < '2019-12-09 00:00:00+00'::timestamp with time zone))

```

But this cost will be only on first time. After that the query and cost would be similar to this, which is still quite high.

```
icds_ucr=> EXPLAIN UPDATE  temp_rohit_activity user_activity
        SET
            last_activity = ut.last_activity
        FROM (
        SELECT audit.username, max(audit.time_of_use)  as last_activity from temp_rohit_activity left join icds_audit_entry_record audit
        oN temp_rohit_activity.username = audit.username
        where audit.time_of_use>'2019-12-06' and audit.time_of_use<'2019-12-09'
        group by audit.username
        )ut                    
        where user_activity.username = ut.username;          
                                                                                                QUERY PLAN                                                                                                 
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Update on temp_rohit_activity user_activity  (cost=87726.95..87837.80 rows=34254 width=113)
   ->  Hash Join  (cost=87726.95..87837.80 rows=34254 width=113)
         Hash Cond: ((user_activity.username)::text = (ut.username)::text)
         ->  Seq Scan on temp_rohit_activity user_activity  (cost=0.00..22.00 rows=1200 width=38)
         ->  Hash  (cost=87655.59..87655.59 rows=5709 width=110)
               ->  Subquery Scan on ut  (cost=87541.41..87655.59 rows=5709 width=110)
                     ->  HashAggregate  (cost=87541.41..87598.50 rows=5709 width=43)
                           Group Key: audit.username
                           ->  Hash Join  (cost=71354.03..86858.45 rows=136592 width=43)
                                 Hash Cond: ((temp_rohit_activity.username)::text = (audit.username)::text)
                                 ->  Seq Scan on temp_rohit_activity  (cost=0.00..22.00 rows=1200 width=32)
                                 ->  Hash  (cost=57519.08..57519.08 rows=649836 width=43)
                                       ->  Append  (cost=0.43..57519.08 rows=649836 width=43)
                                             ->  Index Scan using icds_audit_entry_record_time_of_use_cd233dab on icds_audit_entry_record audit  (cost=0.43..135.20 rows=1464 width=43)
                                                   Index Cond: ((time_of_use > '2019-12-06 00:00:00+00'::timestamp with time zone) AND (time_of_use < '2019-12-09 00:00:00+00'::timestamp with time zone))
                                             ->  Index Scan using icds_audit_entry_record_y2019m12_time_of_use_idx on icds_audit_entry_record_y2019m12 audit_1  (cost=0.43..54134.70 rows=648372 width=43)
                                                   Index Cond: ((time_of_use > '2019-12-06 00:00:00+00'::timestamp with time zone) AND (time_of_use < '2019-12-09 00:00:00+00'::timestamp with time zone))
(17 rows)
```